flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java

```diff
@@ -103,6 +111,12 @@
      */
     private int numBuffersForSort;
 
+    /**
+     * Number of reserved network buffers for data writing. This value can be 0 and 0 means that
+     * {@link HashBasedPartitionSortedBuffer} will be used.
+     */
+    private int numBuffersForWrite;
```

Comment:
From the following modification, it seems the variable here mainly plays a role to indicate whether we want to use hash-based implementation and sort-based implementation. I think perhaps we could directly use a variable like `sortBufferType` or `useHashBuffer` to make it more clear. The `numBuffersForWrite` could be changed to be a local variable in the constructor. We could change the implementation to be like 

```
if (numRequiredBuffer >= 2 * numSubpartitions) {
    useHashBuffer = true;
} else {
    useHashBuffer = false;
}

if (!useHashBuffer) {
    int expectedWriteBuffers;
    if (numRequiredBuffer >= 2 * numSubpartitions) {
        expectedWriteBuffers = 0;
    } else if (networkBufferSize >= NUM_WRITE_BUFFER_BYTES) {
        expectedWriteBuffers = 1;
    } else {
        expectedWriteBuffers =
            Math.min(EXPECTED_WRITE_BATCH_SIZE, NUM_WRITE_BUFFER_BYTES / networkBufferSize);
    }

    int numBuffersForWrite = Math.min(numRequiredBuffer / 2, expectedWriteBuffers);
    numBuffersForSort = numRequiredBuffer - numBuffersForWrite;

    try {
        for (int i = 0; i < numBuffersForWrite; ++i) {
            MemorySegment segment = bufferPool.requestMemorySegmentBlocking();
            writeSegments.add(segment);
        }
    } catch (InterruptedException exception) {
        // the setup method does not allow InterruptedException
        throw new IOException(exception);
    }
}
```